Index: codereview/urls.py
===================================================================
--- codereview/urls.py	(revision 649)
+++ codereview/urls.py	(working copy)
@@ -77,7 +77,7 @@
     (r'^_ah/mail/(.*)', 'incoming_mail'),
     (r'^xsrf_token$', 'xsrf_token'),
     # patching upload.py on the fly
-    (r'^static/upload.py$', 'customized_upload_py'),
+    (r'^dynamic/upload.py$', 'customized_upload_py'),
     (r'^search$', 'search'),
     )
 
